core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -233,6 +233,14 @@ private[spark] class ExecutorAllocationManager(
   }
 
   /**
+   * Change the value of numExecutorsTarget.
+   */
+  def reSetNumExecutorsTarget(): Unit = {
+   logDebug(s"Now reset the value of numExecutorsTarget.")
+    numExecutorsTarget = conf.getInt("spark.dynamicAllocation.initialExecutors", minNumExecutors)
```

Comment:
If I understand correctly what's going on, this will result in `addExecutors` to issue a request for the desired number of executors, if that does not match `spark.dynamicAllocation.initialExecutors`, right? That same method will also bring `numExecutorsTarget` up to whatever was the previous value once it runs.

Instead, wouldn't it be clearer to just explicitly send a message to the AM saying "this is the current state of the world, initialize yourself to deal with it"? Maybe as a reply to the `RegisterClusterManager` message.

Also, `YarnAllocator` holds a lot of state about existing containers, such as in `allocatedHostToContainersMap` and `releasedContainers`. Is that data re-created somehow when the new AM comes up? If not, what are the side-effects, if any, of not having that information around?